crates/extensions/c8y_mapper_ext/src/service_monitor.rs

```diff
@@ -25,47 +26,37 @@ fn default_type() -> String {
     "".to_string()
 }
 
-#[derive(Deserialize, Serialize, Debug)]
-pub struct TopicInfo {
-    pub service_name: String,
-    pub child_id: Option<String>,
-}
-
-impl TopicInfo {
-    fn parse_topic_info(topic: &str) -> Self {
-        let topic_split: Vec<&str> = topic.split('/').collect();
-        let service_name = if topic_split.len() == 4 {
-            topic_split[3]
-        } else {
-            topic_split[2]
-        }
-        .to_string();
-
-        let child_id = if topic_split.len() == 4 {
-            Some(topic_split[2].to_owned())
-        } else {
-            None
-        };
-
-        Self {
-            service_name,
-            child_id,
-        }
-    }
-}
-
 pub fn convert_health_status_message(
+    entity: &EntityTopicId,
     message: &Message,
     device_name: String,
     default_service_type: String,
 ) -> Vec<Message> {
     let mut mqtt_messages: Vec<Message> = Vec::new();
-    let topic = message.topic.name.to_owned();
-    let topic_info = TopicInfo::parse_topic_info(&topic);
+
+    let service_name = entity
+        .default_service_name()
+        .expect("EntityTopicId should be in default scheme");
+
+    let parent = entity
+        .default_parent_identifier()
+        .expect("EntityTopicId should be in default scheme");
+
+    let child_id = if parent.is_default_child_device() {
+        Some(
+            parent
+                .default_device_name()
+                .expect("EntityTopicId should be in default scheme")
+                .to_string(),
+        )
+    } else {
+        None
+    };
```

Comment:
All these information should not be guessed from the entity topic identifier by given by the entity store.
The entity store might assume a default topic scheme (as done here) but only when appropriate, i.e. when no specific information has been provided by registration messages.

This function must be given an `EntityMetadata` and not the `EntityTopicId` of the source.